File: ZezeJava/ZezeJava/src/main/java/Zeze/Game/LoadReporter.java
package Zeze.Game;

import java.util.concurrent.Future;
import Zeze.Builtin.Provider.BLoad;
import Zeze.Net.Binary;
import Zeze.Serialize.ByteBuffer;
import Zeze.Services.ServiceManager.BServerLoad;

public class LoadReporter {
	private long lastLoginTime;
	private int reportDelaySeconds;
	private int timeoutDelaySeconds;
	private Future<?> timerTask;

	public final Online online;

	public LoadReporter(Online online) {
		this.online = online;
	}

	public final void start() {
		start(1);
	}

	public final void start(int delaySeconds) {
		timeoutDelaySeconds = delaySeconds;
		if (null != timerTask)
			timerTask.cancel(false);
		timerTask = Zeze.Util.Task.scheduleUnsafe(timeoutDelaySeconds * 1000L, this::onTimerTask);
	}

	public final void stop() {
		if (null != timerTask) {
			timerTask.cancel(true);
			timerTask = null;
		}
	}

	private void onTimerTask() {
		int online = this.online.getLocalCount();
		long loginTimes = this.online.getLoginTimes();
		int onlineNew = (int)(loginTimes - lastLoginTime);
		lastLoginTime = loginTimes;

		int onlineNewPerSecond = onlineNew / timeoutDelaySeconds;
		//noinspection ConstantConditions
		var config = this.online.providerApp.distribute.loadConfig;
		if (onlineNewPerSecond > config.getMaxOnlineNew()) {
			// 最近上线太多，马上报告负载。linkd不会再分配用户过来。
			report(online, onlineNew);
			// new delay for digestion
			start(onlineNewPerSecond / config.getMaxOnlineNew() + config.getDigestionDelayExSeconds());
			// 消化完后，下一次强迫报告Load。
			reportDelaySeconds = config.getReportDelaySeconds();
			return;
		}
		// slow report
		reportDelaySeconds += timeoutDelaySeconds;
		if (reportDelaySeconds >= config.getReportDelaySeconds()) {
			reportDelaySeconds = 0;
			report(online, onlineNew);
		}
		start();
	}

	public void report(int online, int onlineNew) {
		var load = new BLoad();
		load.setOnline(online);
		//noinspection ConstantConditions
		load.setProposeMaxOnline(this.online.providerApp.distribute.loadConfig.getProposeMaxOnline());
		load.setOnlineNew(onlineNew);
		var bb = ByteBuffer.Allocate(256);
		load.encode(bb);

		var loadServer = new BServerLoad();
		loadServer.ip = this.online.providerApp.directIp;
		loadServer.port = this.online.providerApp.directPort;
		loadServer.param = new Binary(bb);

		this.online.providerApp.zeze.getServiceManagerAgent().setServerLoad(loadServer);
	}
}

File: Zeze/Game/LoadReporter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Zeze.Builtin.Provider;
using Zeze.Net;
using Zeze.Serialize;

namespace Zeze.Game
{
	public class LoadReporter
    {
		private long LastLoginTimes;
		private int ReportDelaySeconds;
		private int TimoutDelaySeconds;
		private Util.SchedulerTask Timer;

		public Online Online { get; }

		public LoadReporter(Online online)
        {
			Online = online;
        }

		public void Start(int delaySeconds = 1)
		{
			TimoutDelaySeconds = delaySeconds;
			Timer?.Cancel();
			Timer = Zeze.Util.Scheduler.Schedule(OnTimerTask, TimoutDelaySeconds * 1000);
		}

		public void Stop()
		{
			Timer?.Cancel();
			Timer = null;
		}

		private void OnTimerTask(Zeze.Util.SchedulerTask ThisTask)
		{
			int online = Online.LocalCount;
			long loginTimes = Online.LoginTimes;
			int onlineNew = (int)(loginTimes - LastLoginTimes);
			LastLoginTimes = loginTimes;

			int onlineNewPerSecond = onlineNew / TimoutDelaySeconds;
			var config = Online.ProviderApp.Distribute.LoadConfig;
			if (onlineNewPerSecond > config.MaxOnlineNew)
			{
				// 最近上线太多，马上报告负载。linkd不会再分配用户过来。
				Report(online, onlineNew);
				// new delay for digestion
				Start(onlineNewPerSecond / config.MaxOnlineNew + config.DigestionDelayExSeconds);
				// 消化完后，下一次强迫报告Load。
				ReportDelaySeconds = config.ReportDelaySeconds;
				return;
			}
			// slow report
			ReportDelaySeconds += TimoutDelaySeconds;
			if (ReportDelaySeconds >= config.ReportDelaySeconds)
			{
				ReportDelaySeconds = 0;
				Report(online, onlineNew);
			}
			Start();
		}

		public void Report(int online, int onlineNew)
		{
			var load = new BLoad();
			load.Online = online;
			var config = Online.ProviderApp.Distribute.LoadConfig;
			load.ProposeMaxOnline = config.ProposeMaxOnline;
			load.OnlineNew = onlineNew;
			var bb = ByteBuffer.Allocate(256);
			load.Encode(bb);

			var loadServer = new Zeze.Services.ServiceManager.ServerLoad();
			loadServer.Ip = Online.ProviderApp.DirectIp;
			loadServer.Port = Online.ProviderApp.DirectPort;
			loadServer.Param = new Binary(bb);

			Online.ProviderApp.Zeze.ServiceManagerAgent.SetServerLoad(loadServer);
		}
	}
}
